### DISCUSSION TITLE
unable to find pylsp

### DISCUSSION THREAD

**Discussion by Obsidianage** [zed-industries/zed#33121] [Feature Requests]
### Check for existing issues

- [ ] Completed

### Describe the bug / provide steps to reproduce it

Language server error: pylsp

failed to spawn command. path: "C:\\Users\\AppData\\Local\\Zed\\languages\\pylsp\\pylsp-venv\\bin\\pylsp", working directory: "D:\\Coding\\Python", args: []
-- stderr--

### Environment

python

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```

```
</details>

**Comment by gryevns** [comment]
Occurs on mac too, can be ignored by disabling pylsp via settings
```
  "languages": {
    "Python": {
      "language_servers": ["!pylsp", "..."]
    }
  }
  ```
  
Running version 0.163.3

**Comment by shakfu** [comment]
I'm getting the same problem on an M1 Mac on Zed version 0.166.1 (01883a4e5394f724c8a65782df74f69b6452b485) where even if add 

```json
"languages": {
      "Python": {
        "language_servers": ["!pylsp"]
      }
    }
```

I get the error:

```text
Language server error: pylsp

oneshot canceled
-- stderr--
~/Library/Application Support/Zed/languages/pylsp/pylsp-venv/bin/pylsp: line 2: ~/Library/Application Support/Zed/languages/pylsp/pylsp-venv/bin/python3.13: No such file or directory
~/Library/Application Support/Zed/languages/pylsp/pylsp-venv/bin/pylsp: line 2: exec: ~/Library/Application Support/Zed/languages/pylsp/pylsp-venv/bin/python3.13: cannot execute: No such file or directory
```

**Comment by shakfu** [comment]
As a fix to the error my prior post, I suggest to remove and re-install the zed editor and then if you still get the error as I did, then do the following:

```sh
cd ~/Library/Application\ Support/Zed/languages/pylsp
rm -rf pylsp-venv
virtualenv pylsp-venv
source pylsp-venv/bin/activate
pip install "python-lsp-server[all]"
```

The pylsp error disappeared after the above.

**Comment by TorratDev** [comment]
Hello everyone! :D

I had the same problem on windows 11 and I made a PR with a simple change to make it work, at least on windows 11.

The PR: #22587

If you try it and it fails let me know and I'll check it again.

**Comment by ZachZimm** [comment]
On Arch Linux I fixed this by installing python-lsp-server from the package manager.

**Comment by aretrace** [comment]
have this issue as well, on macOS

**Comment by Phelsong** [comment]
This is a version mismatch with the pylsp-venv wheels, if you delete it will auto rebuild next launch and resolve itself.
This broke on arch when they moved from 3.12 -> 3.13

![Image](https://github.com/user-attachments/assets/2552db3b-6bb6-4664-95d4-73130aee18fb)

delete:
`~/.local/share/zed/languages/pylsp/pylsp-venv`
or
`$HOME\AppData\Local\Zed\languages\pylsp\pylsp-venv`
or
w/e the mac path is



They probably need to add a check of the versions on startup and rebuild on mismatch.

**Comment by bobrocke** [comment]
> This is a version mismatch with the pylsp-venv wheels, if you delete it will auto rebuild next launch and resolve itself. This broke on arch when they moved from 3.12 -> 3.13
> 
> ![Image](https://github.com/user-attachments/assets/2552db3b-6bb6-4664-95d4-73130aee18fb)
> 
> delete: `~/.local/share/zed/languages/pylsp/pylsp-venv` or `$HOME\AppData\Local\Zed\languages\pylsp\pylsp-venv` or w/e the mac path is
> 

What is the Mac path?

**Comment by hobson** [comment]
I had the same problem on Linux (Ubuntu/Mint). @shakfu 's OSX answer worked for me after changing the application path to `~/.local/share/zed` and the `virtualenv` command to  `python -m virtualenv ...`. This ensure's the correct `virtualenv` and `python` versions on my system. 

> cd ~/.local/share/zed/languages/pylsp
> rm -rf pylsp-venv
> python -m virtualenv pylsp-venv
> source pylsp-venv/bin/activate
> pip install "python-lsp-server[all]"

**Comment by fkarlsson** [comment]
> > This is a version mismatch with the pylsp-venv wheels, if you delete it will auto rebuild next launch and resolve itself. This broke on arch when they moved from 3.12 -> 3.13
> > ![Image](https://github.com/user-attachments/assets/2552db3b-6bb6-4664-95d4-73130aee18fb)
> > delete: `~/.local/share/zed/languages/pylsp/pylsp-venv` or `$HOME\AppData\Local\Zed\languages\pylsp\pylsp-venv` or w/e the mac path is
> 
> What is the Mac path?

The Mac path is here:

`~/Library/Application Support/Zed/languages/pylsp/pylsp-venv`

I faced this issue and did as was suggested by @Phelsong, deleted `pylsp-venv` and restarted Zed. The directory is recreated and I've had no issues after that.

**Comment by devzeth** [comment]
Can reproduce on my mac as well in zed preview 175.3, and deleting as suggesting above worked. But shouldn't zed take care of this @osiewicz ?

**Comment by kkettinger** [comment]
Fresh install of zed `0.174.8` under linux (tuxedo-os/ubuntu):
```
Language server error: pylsp

failed to spawn command. path: "/home/kk/.local/share/zed/languages/pylsp/pylsp-venv/bin/pylsp", working directory: "/home/kk/projects/test args: []
-- stderr--
```

The bin folder doesn't contain `pylsp`:
```
$ ls /home/kk/.local/share/zed/languages/pylsp/pylsp-venv/bin/
insgesamt 0
lrwxrwxrwx 1 kk kk  7 Feb 26 08:42 python -> python3
lrwxrwxrwx 1 kk kk 16 Feb 26 08:42 python3 -> /usr/bin/python3
lrwxrwxrwx 1 kk kk  7 Feb 26 08:42 python3.12 -> python3
```

**Comment by devashish2024** [comment]
After months, I just fixed it for myself atleast! (On Debian latest)

```bash
sudo apt-get install python3-venv
rm ~/.local/share/zed/languages/pylsp/pylsp-venv/ -rf # unsure but just worked
```

> I already had `python3-virtualenv` installed and `virtualenv .venv` works, but still, doing this command and then removing the pylsp-venv folder, then force-closing zed and re-launching it worked!

Source of the command (1): https://askubuntu.com/a/1474910

Also solution for #20028

**Reply by yomiko451** [reply]
fixed it for me on mint22, thanks a lot!

**Comment by tnkuehne** [comment]
```bash
uv pip install python-lsp-server[all]
```
This worked for me on Ubuntu 24.04

**Comment by julienr** [comment]
> After months, I just fixed it for myself atleast! (On Debian latest)
> 
> sudo apt-get install python3-venv
> rm ~/.local/share/zed/languages/pylsp/pylsp-venv/ -rf # unsure but just worked
> 

That fixed it for me as well on Ubuntu 24.04. I was missing the system `python3-venv` package and installing it + forcing zed to create the pylsp-venv fixed the issue.

**Reply by insane-dreamer** [reply]
Fixed issue for me on Ubuntu 24.04 (only deleting the folder under .local and restarting Zed did not resolve it; I had to install python3-venv as well.

**Comment by RomanPro100** [comment]
> After months, I just fixed it for myself atleast! (On Debian latest)
> 
> sudo apt-get install python3-venv
> rm ~/.local/share/zed/languages/pylsp/pylsp-venv/ -rf # unsure but just worked

Fixed it for me, TYSM

**Comment by orazdow** [comment]
If you're using the msys2 windows binary,you have delete the venv folder and rebuild it in native windows, not mysy2:
```
cd %LOCALAPPDATA%\Zed\languages\pylsp
python -m venv pylsp-venv
pylsp-venv\Scripts\activate.bat
pip install "python-lsp-server[all]
```
Zed still says "downloading pylsp" on every start though.. not sure why.

**Comment by stormasm** [comment]
This comment is about running Zed on a Mac.

Deleting Zed and re-installing did NOT work for me and I played around with it for about a day.

The only thing that did work was...

*As noted above* on a Mac this bug can be ignored by disabling pylsp via settings

 ```rust
   "languages": {
     "Python": {
       "language_servers": ["!pylsp", "..."]
     }
   }
```

Running Zed version 0.188.4

**Comment by mikayla-maki** [comment]
It looks like there's a lot of uncertainty about this issue, and the original issue is for a platform we don't support yet. As such, I'm converting this to a discussion so the community has an archive of it :)

**Comment by bepri** [comment]
I encountered this bug on Ubuntu 25.04 with Zed 0.193.1. Running `rm -r ~/.local/share/zed/languages/pylsp/pylsp-venv` and then relaunching Zed appears to have fixed it.

**Comment by SuperKenVery** [comment]
In my case, I 
- Lacked python-venv. Running `sudo apt install python3.8-venv` solved this
- Lacked socks5 support while my network requires it. Editing `pyvenv.cfg` inside the venv and set `include-system-site-packages=true`, and do `sudo apt install python3-socks` solved it.
- After that I need to manually install pylsp: `pip install python-lsp-server[all]`

**Comment by jacksonwb** [comment]
Is there a solution for this on macOS? Currently still getting this on `v0.203.5`. Deleting the venv folder didn't resolve.

**Reply by jacksonwb** [reply]
Ok I resolved by manually activating the environment inside `~/Library/Application Support/Zed/languages/pylsp/pylsp-venv` and installing `python-lsp-server` there.
